File: day17/part1.py
import argparse
import collections
from typing import Callable
from typing import Dict

from support import timing


def run(
        prog: Dict[int, int],
        in_fn: Callable[[], int],
        out_fn: Callable[[int], None],
) -> None:
    def parameter(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[prog[pc + n]]
        elif mode == 1:
            return prog[pc + n]
        elif mode == 2:
            return prog[rb + prog[pc + n]]
        else:
            raise NotImplementedError(mode)

    def store(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[pc + n]
        elif mode == 2:
            return rb + prog[pc + n]
        else:
            raise NotImplementedError(mode)

    rb = 0
    pc = 0
    while pc < len(prog):
        instr = prog[pc]
        opc = instr % 100
        if opc == 99:
            return
        elif opc == 1:
            prog[store(instr, 3)] = parameter(instr, 1) + parameter(instr, 2)
            pc += 4
        elif opc == 2:
            prog[store(instr, 3)] = parameter(instr, 1) * parameter(instr, 2)
            pc += 4
        elif opc == 3:
            prog[store(instr, 1)] = in_fn()
            pc += 2
        elif opc == 4:
            out_fn(parameter(instr, 1))
            pc += 2
        elif opc == 5:
            if parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 6:
            if not parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 7:
            if parameter(instr, 1) < parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 8:
            if parameter(instr, 1) == parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 9:
            rb += parameter(instr, 1)
            pc += 2
        else:
            raise AssertionError(f'unreachable? {prog} {pc}')
    raise AssertionError(f'unreachable? {prog} {pc}')


def compute(s: str) -> int:
    prog = [int(part) for part in s.strip().split(',')]
    prog_d = collections.defaultdict(int, enumerate(prog))

    grid: Dict[int, Dict[int, str]] = collections.defaultdict(lambda: {})

    def in_fn() -> int:
        raise AssertionError('unreachable')

    y = 0
    x = 0

    def out_fn(n: int) -> None:
        nonlocal x
        nonlocal y

        s = chr(n)
        if s == '\n':
            x = 0
            y += 1
        else:
            grid[y][x] = s
            x += 1

    run(prog_d, in_fn, out_fn)

    def is_intersection(y_coord: int, x_coord: int) -> bool:
        return (
            grid[y_coord][x_coord] == '#' and
            y_coord > 0 and
            x_coord > 0 and
            y_coord < len(grid) - 1 and
            x_coord < len(grid[0]) - 1 and
            grid[y_coord - 1][x_coord] == '#' and
            grid[y_coord + 1][x_coord] == '#' and
            grid[y_coord][x_coord + 1] == '#' and
            grid[y_coord][x_coord - 1] == '#'
        )

    ret = 0
    for y_coord, row in grid.items():
        for x_coord, s in row.items():
            if is_intersection(y_coord, x_coord):
                ret += y_coord * x_coord

    return ret


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file')
    args = parser.parse_args()

    with open(args.data_file) as f, timing():
        print(compute(f.read()))

    return 0


if __name__ == '__main__':
    exit(main())

File: day17/part1.php
<?php

if (count($argv) !== 2) {
    echo 'usage: ' . $argv[0] . " FILENAME\n";
}

$prog = json_decode('[' . file_get_contents($argv[1]) . ']', true);

$rb = 0;
$pc = 0;

$x = 0;
$y = 0;
$grid = array(array());

$param = function ($instr, $n, $prog, $rb, $pc) {
    $mode = intdiv($instr, pow(10, ($n + 1))) % 10;
    switch ($mode) {
        case 0:
            return $prog[$prog[$pc + $n] ?? 0] ?? 0;
        case 1:
            return $prog[$pc + $n] ?? 0;
        case 2:
            return $prog[$rb + $prog[$pc + $n] ?? 0] ?? 0;
        default:
            assert("unreachable");
    }
};

$store = function ($instr, $n, $prog, $rb, $pc) {
    $mode = intdiv($instr, pow(10, ($n + 1))) % 10;
    switch ($mode) {
        case 0:
            return $prog[$pc + $n] ?? 0;
        case 2:
            return $rb + $prog[$pc + $n] ?? 0;
        default:
            assert("unreachable");
    }
};

while (TRUE) {
    $instr = $prog[$pc];
    $opc = $instr % 100;
    switch ($opc) {
        case 99:
            break 2;
        case 1:
            $prog[$store($instr, 3, $prog, $rb, $pc)] = $param($instr, 1, $prog, $rb, $pc) + $param($instr, 2, $prog, $rb, $pc);
            $pc += 4;
            break;
        case 2:
            $prog[$store($instr, 3, $prog, $rb, $pc)] = $param($instr, 1, $prog, $rb, $pc) * $param($instr, 2, $prog, $rb, $pc);
            $pc += 4;
            break;
        case 3:
            // TODO input
            $prog[$store($instr, 1, $prog, $rb, $pc)] = 2;
            $pc += 2;
            break;
        case 4:
            // TODO output
            $s = chr($param($instr, 1, $prog, $rb, $pc));
            if ($s === "\n") {
                $x = 0;
                $y += 1;
                $grid[$y] = array();
            } else {
                $grid[$y][$x] = $s;
                $x += 1;
            }
            echo $s;
            $pc += 2;
            break;
        case 5:
            if ($param($instr, 1, $prog, $rb, $pc) !== 0) {
                $pc = $param($instr, 2, $prog, $rb, $pc);
            } else {
                $pc += 3;
            }
            break;
        case 6:
            if ($param($instr, 1, $prog, $rb, $pc) === 0) {
                $pc = $param($instr, 2, $prog, $rb, $pc);
            } else {
                $pc += 3;
            }
            break;
        case 7:
            if ($param($instr, 1, $prog, $rb, $pc) < $param($instr, 2, $prog, $rb, $pc)) {
                $prog[$store($instr, 3, $prog, $rb, $pc)] = 1;
            } else {
                $prog[$store($instr, 3, $prog, $rb, $pc)] = 0;
            }
            $pc += 4;
            break;
        case 8:
            if ($param($instr, 1, $prog, $rb, $pc) === $param($instr, 2, $prog, $rb, $pc)) {
                $prog[$store($instr, 3, $prog, $rb, $pc)] = 1;
            } else {
                $prog[$store($instr, 3, $prog, $rb, $pc)] = 0;
            }
            $pc += 4;
            break;
        case 9:
            $rb += $param($instr, 1, $prog, $rb, $pc);
            $pc += 2;
        default:
            assert('unreachable');
    }
}

// we add an extra empty array at the end
unset($grid[count($grid) - 1]);
// it prints an extra newline at the end
unset($grid[count($grid) - 1]);

$is_intersection = function($grid, $y_coord, $x_coord) {
    return (
        $grid[$y_coord][$x_coord] === '#' &&
        $y_coord > 0 &&
        $x_coord > 0 &&
        $y_coord < count($grid) - 1 &&
        $x_coord < count($grid[0]) - 1 &&
        $grid[$y_coord + 1][$x_coord] === '#' &&
        $grid[$y_coord - 1][$x_coord] === '#' &&
        $grid[$y_coord][$x_coord + 1] === '#' &&
        $grid[$y_coord][$x_coord - 1] === '#'
    );
};

$ret = 0;
foreach ($grid as $y_coord => $row) {
    foreach ($row as $x_coord => $s) {
        if ($is_intersection($grid, $y_coord, $x_coord)) {
            $ret += $y_coord * $x_coord;
        }
    }
}
echo $ret . "\n";
